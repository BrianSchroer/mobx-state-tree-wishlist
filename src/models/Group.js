import { types } from 'mobx-state-tree';
import { User } from './User';

export const Group = types.model({
  users: types.map(User)
});
