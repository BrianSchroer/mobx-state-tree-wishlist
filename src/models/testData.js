export const testGroupInput = {
  users: {
    a342: {
      id: 'a342',
      name: 'Homer',
      gender: 'm'
    },
    '5fc2': {
      id: '5fc2',
      name: 'Marge',
      gender: 'f'
    },
    '663b': {
      id: '663b',
      name: 'Bart',
      gender: 'm'
    },
    '65aa': {
      id: '65aa',
      name: 'Maggie',
      gender: 'f'
    },
    ba32: {
      id: 'ba32',
      name: 'Lisa',
      gender: 'f'
    }
  }
};

export const testWishListItemInput = {
  name: 'Chronicles of Narnia Box Set - C.S. Lewis',
  price: 28.83,
  image:
    'https://images-na.ssl-images-amazon.com/images/I/51LmtX5KPAL._SX406_BO1,204,203,200_.jpg'
};

export const testWishListItemInputs = [1, 2, 3].map(i => ({
  name: `name${i}`,
  price: i,
  image: `image${i}`
}));

export const testUserInput = {
  id: '1',
  name: 'Brian',
  gender: 'm'
};

export const testUserInputWithWishList = {
  id: '1',
  name: 'Brian',
  gender: 'm',
  wishList: {
    items: testWishListItemInputs
  }
};
