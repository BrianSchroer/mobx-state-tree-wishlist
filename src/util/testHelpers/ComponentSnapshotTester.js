import React from 'react';
import renderer from 'react-test-renderer';

/**
 * React component snapshot helper
 */
export default class ComponentSnapshotTester {

  /**
   * Create new ComponentSnapshotTester instance
   * @param {*} component - React component
   */
  constructor(component) {
    this.component = component.type;
    this.props = component.props;
    this.adjustProps = (props, propOverrides) => Object.assign({}, props, propOverrides);
  }

  /**
   * Specify callback function to be called by the test function to adjust component
   * props with propOverrides.
   * (Defaults to adjusting top-level props.)
   */
  withPropAdjustor = (adjustor) => {
    this.adjustProps = adjustor;
    return this;
  };

  /**
   * Adjust props that were passed to the constructor with propOverrides, then create
   * snapshot and assert that it matches previously accepted screenshot.
   */
  test = (propOverrides) => {
    const Component = this.component;
    const props = this.adjustProps(this.props, propOverrides);

    const newSnapshot = renderer.create(<Component {...props} />).toJSON();

    expect(newSnapshot).toMatchSnapshot();
  }

  /**
   * Create snapshot and assert that it matches previously accepted screenshot. 
   * @param {*} component - React component1 
   */
  static test(component) {
    new ComponentSnapshotTester(component).test();
  }
}