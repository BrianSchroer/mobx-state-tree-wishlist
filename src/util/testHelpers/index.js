export { default as mobxSnapshotHelper } from './mobxSnapshotHelper/snapshotHelper';
export { default as ComponentSnapshotHelper } from './ComponentSnapshotHelper/SnapshotHelper';