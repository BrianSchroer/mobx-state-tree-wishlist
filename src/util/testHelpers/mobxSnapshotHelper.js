import { getSnapshot } from 'mobx-state-tree';

export default class {
  static createSnapshot(target) {
    return getSnapshot(target);
  }

  static assertMatch(target) {
    expect(this.createSnapshot(target)).toMatchSnapshot();
  }
}