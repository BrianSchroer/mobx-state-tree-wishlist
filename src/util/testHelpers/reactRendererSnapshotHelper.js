import renderer from 'react-test-renderer';

/**
 * react-test-renderer snapshot helper methods
 */
export default class {

  /**
   * Create component from JSX
   * @param {JSX} componentJsx (e.g. '<Button />')
   * @returns {object} component object
   */
  static createComponent(componentJsx) {
    return renderer.create(componentJsx);
  }

  /**
   * Create component from JSX and serialize as JSON "snapshot" string.
   * @param {JSX} componentJsx (e.g. '<Button />')
   * @returns {string} JSON string.
   */
  static createSnapshot(componentJsx) {
    return this.createComponent(componentJsx).toJSON();
  }

  /**
   * Create component from JSX, serialize as JSON string, and compare to accepted snapshot.
   * If a snapshot doesn't already exist, one is saved. If a previous snapshot exists
   * and the new snapshot doesn't match, you'll be asked if that's an error or if you want
   * to accept the new snapshot.
   * @param {JSX} componentJsx (e.g. '<Button />')
   */
  static assertMatch(componentJsx) {
    expect(this.createSnapshot(componentJsx)).toMatchSnapshot();
  }
}