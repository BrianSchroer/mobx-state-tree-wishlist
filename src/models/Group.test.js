import { Group } from './Group';
import { testGroupInput } from './testData';
import { mobxSnapshotHelper } from '../util/testHelpers';

describe('Group', () => {
  it('.create() should return expected items', () => {
    mobxSnapshotHelper.test(Group.create(testGroupInput));
  });
});
