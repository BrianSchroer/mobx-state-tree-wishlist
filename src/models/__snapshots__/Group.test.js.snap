// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Group .create() should return expected items 1`] = `
Object {
  "users": Object {
    "5fc2": Object {
      "gender": "f",
      "id": "5fc2",
      "name": "Marge",
      "wishList": Object {
        "items": Array [],
      },
    },
    "65aa": Object {
      "gender": "f",
      "id": "65aa",
      "name": "Maggie",
      "wishList": Object {
        "items": Array [],
      },
    },
    "663b": Object {
      "gender": "m",
      "id": "663b",
      "name": "Bart",
      "wishList": Object {
        "items": Array [],
      },
    },
    "a342": Object {
      "gender": "m",
      "id": "a342",
      "name": "Homer",
      "wishList": Object {
        "items": Array [],
      },
    },
    "ba32": Object {
      "gender": "f",
      "id": "ba32",
      "name": "Lisa",
      "wishList": Object {
        "items": Array [],
      },
    },
  },
}
`;
